modules/AlphaWalletFoundation/AlphaWalletFoundation/Transfer/eip1559/SuggestEip1559.swift

```diff
@@ -0,0 +1,220 @@
+                    block: .blockNumber(value: lastBlock),
+                    rewardPercentile: rewardPercentile).first
+
+                let rewardSize = feeHistory.reward.count
+                guard !feeHistory.reward.isEmpty else { throw EIP1559Error.rewardsNotFound }
+                print("[GasEstimator] rewards raw: \(feeHistory.reward)")
+                for reward in feeHistory.reward {
+                    let value = reward[0]
+                    rewards += [value]
+                }
+                if rewardSize < blockCount {
+                    break
+                }
+                needBlocks -= blockCount
+            }
+            ptr -= blockCount + 1
+        }
+
+        if rewards.isEmpty {
+            return fallbackPriorityFee
+        }
+        rewards.sort()
+
+        let index = Int(floor(Double((rewards.count - 1) * rewardBlockPercentile) / 100.0))
+
+        print("[GasEstimator] rewards: \(rewards)")
+        print("[GasEstimator] priorityFee: \(rewards[index])")
+        print("[GasEstimator] priorityFee index: \(index)")
+
+        return Decimal(rewards[index])
+    }
+
+    enum EIP1559Error: Error {
+        case baseFeeIsEmpty
+        case firstBlockMissing
+        case failureToBuildTimedEstimate
+        case indexedRewardNotFound
+        case rewardsNotFound
+        case indexedBaseFeeNotFound
+        case minBaseFeeNotFound
+    }
+
+    func calculateResult(priorityFee: Decimal, feeHistory: FeeHistory) throws -> [Int: EIP1559FeeOracleResult] {
+        var baseFee = feeHistory.baseFeePerGas
+
+        print("[GasEstimator] baseFee.-1: \(feeHistory.baseFeePerGas)")
+        print("[GasEstimator] baseFee.0: \(baseFee)")
+
+        guard !baseFee.isEmpty else { throw EIP1559Error.baseFeeIsEmpty }
+
+        print("[GasEstimator] last in base fee.0: \(baseFee[baseFee.count - 1])")
+
+        let lastBaseFee = baseFee[baseFee.count - 1] * 1.125
+        baseFee[baseFee.count - 1] = lastBaseFee
+        print("[GasEstimator] last value: \(lastBaseFee)")
+
+        for i in feeHistory.gasUsedRatio.indices.reversed() where feeHistory.gasUsedRatio[i] > 0.9 {
+            baseFee[i] = baseFee[i + 1]
+        }
+
+        let order: [Int] = Array(0 ... feeHistory.gasUsedRatio.count).indices.sorted {
+            guard let first = baseFee[safe: $0] else { return false }
+            guard let second = baseFee[safe: $1] else { return true }
+            return first < second
+        }
+
+        print("[GasEstimator] order: \(order)")
+
+        var maxBaseFee: Decimal = .zero
+        var result: [Int: EIP1559FeeOracleResult] = [:]
+
+        print("[GasEstimator] baseFee.1: \(baseFee)")
+        print("[GasEstimator] timeFactors: \((0...maxTimeFactor).reversed())")
+
+        for timeFactor in (0...maxTimeFactor).reversed() {
+            var bf = try predictMinBaseFee(baseFee: baseFee, order: order, timeFactor: timeFactor)
+            var t = priorityFee
+            bf = max(bf, maxBaseFee)
```

Comment:
This will cause the next line`if bf > maxBaseFee` always be false, right? Think have to tweak it a bit so the if-else is there and the `else` sometimes executes.

Also not sure if it should be `max()` or `min()` (and the `maxBaseFee` name accordingly). Can you help verify again?

Reply:
> if bf > maxBaseFee it looks like no, i am able to enter here

Oh, you are right.

> updating code with min function returns incorrect values for me

And yes, looks like bf ought to increase (or stay unchanged) with each iteration in the `for-each`, so max is correct.

Thanks for checking :)

Reply:
> @hboon do we have any validation rules for `priorityFee`? except safe range `0...4`

I don't know of any.

Reply:
But because of the max() this is always 0, right?

```
(maxBaseFee - bf) * extraPriorityFeeRatio
```

So `t` is the same for all the estimates.

Reply:
Anyway, if you'd like, we can merge this and keep the Feature flag disabled, then work out the estimation? Then less work rebasing to keep the branch up to date.

Reply:
> So t is the same for all the estimates.

yes